# Проект по SQL

Проанализировать базу данных крупного сервиса для чтения книг по подписке.

## Описание данных

### Таблица `books`

Содержит информацию о книгах:

`book_id` — id книги;

`author_id` — id автора;

`title` — название книги;

`num_pages` — количество страниц;

`publication_date` — дата выпуска книги;

`publisher_id` — id издателя.


### Таблица `authors`

Содержит информацию об авторах:

`author_id` — id автора;

`author` — имя автора.

### Таблица `publishers`

Содержит информацию об издательствах:

`publisher_id` — id издательства;

`publisher` — название издательства;

### Таблица `ratings`

Содержит информацию о пользовательских оценках книг:

`rating_id` — id оценки;

`book_id` — id книги;

`username` — имя пользователя, оставившего оценку;

`rating` — рейтинг книги.

### Таблица `reviews`

Содержит информацию о пользовательских рецензиях:

`review_id` — id рецензии;

`book_id` — id книги;

`username` — имя автора рецензии;

`text` — текст рецензии.






# План исследования <a class="anchor" id="0-bullet">
* [Цели исследования](#1-bullet)
* [Вывод первых строк](#2-bullet)
* [SQL запросы](#3-bullet)
    * [Количество книг вышедших после 1 января 2000 года](#4-bullet)
    * [Количество обзоров и средняя оценка для каждой книги](#5-bullet)
    * [Издательство, которое выпустило наибольшее число книг толще 50 страниц](#6-bullet)
    * [Автор с самой высокой средней оценкой книг. Книги с 50 и более оценками](#7-bullet)
    * [Среднее количество обзоров от пользователей, которые поставили больше 50 оценок](#8-bullet)

### Цели исследования<a class="anchor" id="1-bullet"></a>
👈[назад к оглавлению](#0-bullet)

Проанализировать базу данных крупного сервиса для чтения книг по подписке.


### Вывод первых строк<a class="anchor" id="2-bullet"></a>
👈[назад к оглавлению](#0-bullet)

Подключение к базе данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
from IPython.display import display

# устанавливаем параметры
db_config = {'user': '123', # имя пользователя
 'pwd': '123', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 


In [2]:
query = '''
           SELECT *
           FROM books
        '''
 
books = pd.io.sql.read_sql(query, con = engine)
 
books.to_csv('books.csv', index = False)
display(books.head())
books.info()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
book_id             1000 non-null int64
author_id           1000 non-null int64
title               1000 non-null object
num_pages           1000 non-null int64
publication_date    1000 non-null object
publisher_id        1000 non-null int64
dtypes: int64(4), object(2)
memory usage: 47.0+ KB


In [3]:
query = '''
           SELECT *
           FROM authors
        '''
 
authors = pd.io.sql.read_sql(query, con = engine)
 
authors.to_csv('authors.csv', index = False)
display(authors.head())
authors.info()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
author_id    636 non-null int64
author       636 non-null object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


In [4]:
query = '''
           SELECT *
           FROM publishers
        '''
 
publishers = pd.io.sql.read_sql(query, con = engine)
 
publishers.to_csv('publishers.csv', index = False)
display(publishers.head())
publishers.info()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 340 entries, 0 to 339
Data columns (total 2 columns):
publisher_id    340 non-null int64
publisher       340 non-null object
dtypes: int64(1), object(1)
memory usage: 5.4+ KB


In [5]:
query = '''
           SELECT *
           FROM ratings
        '''
 
ratings = pd.io.sql.read_sql(query, con = engine)
 
ratings.to_csv('ratings.csv', index = False)
display(ratings.head())
ratings.info()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
rating_id    6456 non-null int64
book_id      6456 non-null int64
username     6456 non-null object
rating       6456 non-null int64
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


In [6]:
query = '''
           SELECT *
           FROM reviews
        '''
 
reviews = pd.io.sql.read_sql(query, con = engine)
 
reviews.to_csv('reviews.csv', index = False)
display(reviews.head())
reviews.info()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
review_id    2793 non-null int64
book_id      2793 non-null int64
username     2793 non-null object
text         2793 non-null object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


### SQL запросы<a class="anchor" id="3-bullet"></a>
👈[назад к оглавлению](#0-bullet)

### Количество книг вышедших после 1 января 2000 года<a class="anchor" id="4-bullet"></a>
👈[назад к оглавлению](#0-bullet)

In [7]:
query = '''
           SELECT
               COUNT(*) AS cnt 
           FROM 
               books
           WHERE 
               publication_date >= '2000-01-01';
        '''
 
req_1 = pd.io.sql.read_sql(query, con = engine)
req_1

cnt
0  821

###### Вывод

После 1 января 2000 года вышла 821 книга.

### Количество обзоров и средняя оценка для каждой книги<a class="anchor" id="5-bullet"></a>
👈[назад к оглавлению](#0-bullet)

In [8]:
query = '''
           SELECT
               books.book_id,
               title,
               AVG(rating) AS avg_rating,
               sub.count_reviews
           FROM
               books
           INNER JOIN ratings ON ratings.book_id = books.book_id
            
           INNER JOIN
                    (SELECT
                         book_id,
                         COUNT(text) AS count_reviews
                     FROM
                         reviews
                     GROUP BY
                         book_id
                         
                         ) AS Sub             
                      
           ON Sub.book_id = books.book_id
           
           GROUP BY 
               books.book_id,
               books.title,
               sub.count_reviews
           ORDER BY
               book_id ASC;

        '''
 
req_2_1 = pd.io.sql.read_sql(query, con = engine) 
req_2_1.head()

book_id                                              title  avg_rating  \
0        1                                       'Salem's Lot    3.666667   
1        2                 1 000 Places to See Before You Die    2.500000   
2        3  13 Little Blue Envelopes (Little Blue Envelope...    4.666667   
3        4  1491: New Revelations of the Americas Before C...    4.500000   
4        5                                               1776    4.000000   

   count_reviews  
0              2  
1              1  
2              3  
3              2  
4              4

###### Вывод

`avg_rating` - средняя оценка, `cnt` - количество обзоров.

### Издательство, которое выпустило наибольшее число книг толще 50 страниц<a class="anchor" id="6-bullet"></a>
👈[назад к оглавлению](#0-bullet)

In [9]:
query = '''
           SELECT
               books.publisher_id AS publisher_id,
               publishers.publisher AS publisher,
               COUNT(book_id) AS cnt               
           FROM 
               books
           INNER JOIN publishers ON publishers.publisher_id = books.publisher_id
    
           WHERE 
               num_pages >= 50
           GROUP BY
               books.publisher_id,
               publishers.publisher           
           ORDER BY
               cnt DESC
           LIMIT 5;
           
        '''
 
req_3 = pd.io.sql.read_sql(query, con = engine)
req_3

publisher_id                 publisher  cnt
0           212             Penguin Books   42
1           309                   Vintage   31
2           116  Grand Central Publishing   25
3           217          Penguin Classics   24
4            35                    Bantam   19

###### Вывод

Издательство Penguin Books выпустило наибольшее число книг толще 50 страниц - 42 шт.

### Автор с самой высокой средней оценкой книг. Книги с 50 и более оценками<a class="anchor" id="7-bullet"></a>
👈[назад к оглавлению](#0-bullet)

In [10]:
query = '''
           SELECT
               books.author_id AS author_id,
               authors.author AS author,
               AVG(SUBQ.AVG_rating) AS avg_by_author
           FROM 
               books
           INNER JOIN authors ON authors.author_id = books.author_id
           INNER JOIN
               (
                SELECT
                    book_id,
                    AVG_rating
                FROM
                    (SELECT
                         book_id,
                         AVG(rating) AS AVG_rating,
                         COUNT(rating) as count_rating
                     FROM 
                         ratings
                     GROUP BY
                         book_id) AS Sub 
                WHERE
                    SUB.count_rating > 50    
               
                GROUP BY
                    Sub.book_id,
                    AVG_rating


               ) AS SUBQ
           ON 
               SUBQ.book_id = books.book_id
           
           
           GROUP BY
               books.author_id,
               authors.author
           ORDER BY
               avg_by_author DESC;
           
        '''
 
req_4_1 = pd.io.sql.read_sql(query, con = engine)
req_4_1

author_id                                             author  \
0         236                         J.K. Rowling/Mary GrandPré   
1         402                  Markus Zusak/Cao Xuân Việt Khương   
2         240                                     J.R.R. Tolkien   
3         376                                  Louisa May Alcott   
4         498                                       Rick Riordan   
5         621                                    William Golding   
6         235                                      J.D. Salinger   
7         469           Paulo Coelho/Alan R. Clarke/Özdemir İnce   
8         630  William Shakespeare/Paul Werstine/Barbara A. M...   
9         106                                          Dan Brown   
10        372                                         Lois Lowry   
11        195            George Orwell/Boris Grabnar/Peter Škerl   
12        554                                    Stephenie Meyer   
13        311                                     John Steinbeck   

    avg_by_author  
0        4.283844  
1        4.264151  
2        4.258446  
3        4.192308  
4        4.080645  
5        3.901408  
6        3.825581  
7        3.789474  
8        3.787879  
9        3.754540  
10       3.750000  
11       3.729730  
12       3.662500  
13       3.622951

###### Вывод

Автор с самой высокой средней оценкой книг - J.K. Rowling/Mary GrandPré. Cредний рейтинг автора по всем его книгам - 4,28.

### Среднее количество обзоров от пользователей, которые поставили больше 50 оценок<a class="anchor" id="8-bullet"></a>
👈[назад к оглавлению](#0-bullet)

In [11]:
query = '''
           SELECT
               AVG(Subq1.count_reviews) AS avg_reviews
           FROM 
           
               (SELECT
                    username,
                    COUNT(reviews.review_id) as count_reviews
                FROM 
                    reviews
                GROUP BY
                    reviews.username
                ) AS Subq1
           INNER JOIN
               (
                SELECT
                    username,
                    COUNT(ratings.rating) AS cnt
                FROM
                    ratings 
                GROUP BY
                    ratings.username

               ) AS SUBQ2
           ON 
               SUBQ2.username = Subq1.username
           WHERE
               SUBQ2.cnt > 50;
           
        '''
 
req_5_1 = pd.io.sql.read_sql(query, con = engine)
req_5_1

avg_reviews
0    24.333333

###### Вывод

Среднее количество обзоров - 24,3.
